File: haskell/02-arithmetic/P32.hs
module P32 where
import P31
-- (**) Determine the prime factors of a given positive integer.

--primeFactors :: Int -> [Int]
primeFactors number = concat (filter (\xs -> (length xs) > 0) [(factors number n) | n <- [2..(quot number 2)], isPrime n])

-- factors :: Int -> Int -> [Int]
factors number f
        | number `mod` f == 0 = f : (factors (quot number f) f)
factors _ f = []

File: java8/src/main/java/com/shekhargulati/ninetynine_problems/java8/_02_arithemetic/P32.java
package com.shekhargulati.ninetynine_problems.java8._02_arithemetic;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.IntStream;

/**
 * (**) Determine the prime factors of a given positive integer.
 */
public class P32 {

    public static List<Integer> primeFactors(int number) {
        List<Integer> fs = new ArrayList<>();
        IntStream.rangeClosed(2, number / 2).filter(P31::isPrime).forEach(f -> {
            int n = number;
            while (n % f == 0) {
                n = n / f;
                fs.add(f);
            }
        });
        return fs;
    }
}
